### PR TITLE
Improvements sourced from several shield add-on repositories

### PR DESCRIPTION
**Pull request by motin**
Many improvements to the template. Primarily sourced from:
* https://github.com/motin/taar-experiment-v2-shield-study/
* https://github.com/motin/esper-pioneer-shield-study/
* https://github.com/Dexterp37/icq-shield-study
* https://github.com/biancadanforth/tracking-protection-shield-study

Version bump to 1.3.0

The idea of this PR is to, in one big swoop, bring this template forward to reflect today's improvements and desired changes discovered while developing the above add-ons, as well as make it easier (via npm run format addition and linting improvements) to merge and backport code between add-ons and the template (with less differences only from code style and whitespace).

### CODE REVIEW

**Comment by motin** [comment]
This PR currently includes commits from several open PRs in this repo. They can be approved in this order: First #37 then #41. Don't approve #35 (since it's change is already included in #41). Approve #39 and #30 in any order. 
After that, I'll rebase this PR on top of the current master.

**Comment by gregglind** [comment]
Thanks for these.  I am PTO until Tuesday.

File: .eslintrc.js [hunk 1]
```diff
@@ -26,23 +27,20 @@ module.exports = {
...
 
   rules: {
     "babel/new-cap": "off",
     "comma-dangle": ["error", "always-multiline"],
-    "eqeqeq": "error",
-    "indent": ["warn", 2, {SwitchCase: 1}],
+    eqeqeq: "error",
+    indent: ["warn", 2, { SwitchCase: 1 }],
     "mozilla/no-aArgs": "warn",
     "mozilla/balanced-listeners": 0,
```
Comment (pdehaan) [inline .eslintrc.js]: nit: this should be `"off"`, for consistency.

File: .eslintrc.js [hunk 2]
```diff
@@ -26,23 +27,20 @@ module.exports = {
...
     "no-console": "warn",
     "no-shadow": ["error"],
```
Comment (pdehaan) [inline .eslintrc.js]: nit: this could just be `"error"`, and doesn't need to be wrapped in an array.

File: DEV.md [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+    f.  Feature starts using the `variation` from that info.
+    g.  Feature instruments user button to send `telemetry` and to `endStudy` if the button is clicked enough.
+
+Tip: For more insight on what is study-specific, compare the source code of previously deployed shield studies with this template (and each other) to get an idea of what is actually relevant to change between studies vs what is mostly untouched boilerplate.
+
+### Legacy repositories
+
+Repositories that should no longer be used as templates for new studies:
+
+[https://github.com/gregglind/template-shield-study]() - The incubation repo for the updated structure and contents of this repo, implemented in late 2017.
```
Comment (pdehaan) [inline DEV.md]: This looks "off". Not sure if we should duplicate the URL in the parens, or just use the `<https://github.com/gregglind/template-shield-study>` syntax since the URL and label are the same.

Same w/ the hello world example link below.

File: README.md [hunk 4]
```diff
@@ -2,251 +2,32 @@
...
-
-- a notification bar
-
-    -  introducing the feature.
-    -  allowing them to opt out
-
-During FIRST INSTALL and EVERY OTHER STARTUP, users see:
-
-- a 'toolbar button' (webExtension BrowserAction)
+**Note**: This is toy / demonstration [Shield Study](https://wiki.mozilla.org/Firefox/Shield/Shield_Studies) Legacy Addon. Use this as a template for yours
```
Comment (pdehaan) [inline README.md]: nit "Add-on", for consistency.

File: TESTPLAN.md [hunk 5]
```diff
@@ -1,161 +1,130 @@
...
-        - ending is `voted`
-        - 'vote' is correct.
+2. UI functionality: VOTE
 
-3.  UI functionality: 'X' button
+   Expect: Click on a 'vote' button (any of: `yes | not sure | no`) has all these effects
 
-    Click on the 'x' button.
+   * notice closes
+   * addon uninstalls
```
Comment (pdehaan) [inline TESTPLAN.md]: add-on

File: TESTPLAN.md [hunk 6]
```diff
@@ -1,161 +1,130 @@
...
-      - ending is `notification-x`
+3. UI functionality: 'X' button
 
-4.  UI functionality  'close window'
+   Click on the 'x' button.
 
-    1.  Open a 2nd firefox window.
-    2.  Close the initial window.
+   * notice closes
+   * addon uninstalls
```
Comment (pdehaan) [inline TESTPLAN.md]: nit: add-on

File: TESTPLAN.md [hunk 7]
```diff
@@ -1,161 +1,130 @@
...
-      - ending is `window-or-fx-closed`
+   1. Open a 2nd Firefox window.
+   2. Close the initial window.
 
+   Then observe:
 
----
-## Helper code and tips
+   * notice closes
+   * addon uninstalls
```
Comment (pdehaan) [inline TESTPLAN.md]: nit: add-on

File: TESTPLAN.md [hunk 8]
```diff
@@ -1,161 +1,130 @@
...
-  )
-
-  pings.forEach(p=>{
-    console.log(p.creationDate, p.payload.type);
-    console.log(JSON.stringify(p.payload.data,null,2))
-  })
-}
-
-printPings()
+Example log output after installing the addon:
```
Comment (pdehaan) [inline TESTPLAN.md]: nit: add-on (probably worth a global search and see what needs to be updated).

**Comment by motin** [comment]
Btw, this PR also fixes #38

File: addon/bootstrap.js [hunk 9]
```diff
@@ -62,14 +62,14 @@ async function startup(addonData, reason) {
...
-    * - note first seen,
-    * - check eligible
-    */
-  if ((REASONS[reason]) === "ADDON_INSTALL") {
-    //  telemetry "enter" ONCE
+  /** addon_install and addon_upgrade (which is considered a new study) ONLY:
+   * - note first seen,
+   * - check eligible
+   */
+  if (reason === REASONS.ADDON_INSTALL || reason === REASONS.ADDON_UPGRADE) {
```
Comment (biancadanforth) [inline addon/bootstrap.js]: I can see how this would be useful for testing, but what about in deployment? @motin, @gregglind , If we have already deployed an addon, and we have to patch it, is that an upgrade or a fresh install?

If we do upgrade addons after deployment, do we want the addon to completely reset? I.e. there would multiple `firstSeen` and other initial pings from the old version and the user might see an "install" UI element more than once.

This point is moot if we never see the `ADDON_UPGRADE` reason when the study is re-deployed through Normandy.

**Comment by gregglind** [comment]
It varies by addon, alas, and the author has to be aware of it.  Normandy
tends to do new installs that pave over the old one.  I don’t know if that
is what they *desire* to do.

**Comment by mythmon** [comment]
The add-on manager calls something an upgrade when an add-on with the same ID as an existing add-on is installed, and a greater or equal version

Normandy generally doesn't trigger upgrades, because Normandy doesn't change studies in practice. Instead it installs the add-on once, and then doesn't touch it until unenrollment time. In the future, when Normandy supports changing add-ons in place, it will trigger upgrades.

File: package.json [hunk 10]
```diff
@@ -20,6 +20,9 @@
   "bugs": {
     "url": "https://github.com/mozilla/shield-studies-addon-template/issues"
   },
+  "engines": {
+    "node": ">=8.9.0"
```
Comment (biancadanforth) [inline package.json]: I know we [require](https://github.com/mozilla/shield-studies-addon-template#testing) Node 7.6+ for async/await, but is there a specific reason why we require >= 8.9.0? I read that it's the LTS for Node v8. Just curious mostly.

File: test/utils.js [hunk 11]
```diff
@@ -33,6 +33,9 @@ const FIREFOX_PREFERENCES = {
   // NECESSARY for all 57+ builds
   "extensions.legacy.enabled": true,
 
+  // Include log output in browser console
+  "shield.testing.logging.level": 10, // Trace
```
Comment (biancadanforth) [inline test/utils.js]: I'm actually in favor of a pref in `utils.js`, since then we wouldn't have to remember to turn off logging in `Config.jsm` when shipping the addon. SSAU also pulls the logging level from Config.jsm, so we'd need a patch for it to read the pref instead as well.

Can you update this PR to read the logging level for `initLog()` from this pref in `bootstrap.js` and in any other files like `Feature.jsm` as well?

File: addon/bootstrap.js (hunk 9 above)
Comment (biancadanforth) [reply addon/bootstrap.js]: Based on @mythmon's comments:
> The add-on manager calls something an upgrade when an add-on with the same ID as an existing add-on is installed, and a greater or equal version

> Normandy generally doesn't trigger upgrades, because Normandy doesn't change studies in practice. Instead it installs the add-on once, and then doesn't touch it until unenrollment time. In the future, when Normandy supports changing add-ons in place, it will trigger upgrades.

In light of this, my vote is that we future-proof the template and not equate `ADDON_INSTALL` with `ADDON_UPGRADE`. QA may have to remove the old version before re-installing a new version, but that seems a small price to pay to avoid potential unintended consequences of studies resetting on upgrade in the future. I think what happens on `ADDON_UPGRADE` will vary by study. @mythmon , @gregglind , @raymak, @motin , any feelings one way or the other?

**Comment by gregglind** [comment]
Install and Upgrade are different, and the behaviours for each should be
described / differ by study. Do not equate them.

**Comment by motin** [comment]
Rebased upon latest master. Addressed open concerns via 815ff06 and 9d7fdaa

**Comment by motin** [comment]
@biancadanforth I removed the configuration of study-specific logging since this should not belong in the template to begin with. We should expect a studyUtils.log API, like pioneerUtils has. Good target for https://github.com/mozilla/shield-studies-addon-utils/issues/101

Lazy-loading StudyUtils makes it unavailable outside startup(), this affects when and where logging and REASONS is available. @gregglind Do we have to do a straight Cu.import to make this available in all of bootstrap.js?

**Comment by biancadanforth** [comment]
@motin That's a great point about the logging. Could you add a link to the pioneerUtils as a comment in that issue as a great example of how to do that?

File: addon/bootstrap.js [hunk 12]
```diff
@@ -105,8 +108,8 @@ async function startup(addonData, reason) {
   // log what the study variation and other info is.
   log.debug(`info ${JSON.stringify(studyUtils.info())}`);
 
-  // Start up your feature, with specific variation info.
-  this.feature = new Feature({variation, studyUtils, reasonName: REASONS[reason]});
+  // start up the chrome-privileged part of the study
+  this.feature.privilegedStartup();
```
Comment (biancadanforth) [inline addon/bootstrap.js]: Just curious, why is this better than instantiating the feature once and having its constructor method startup the feature?

File: package.json [hunk 13]
```diff
@@ -45,7 +45,6 @@
     "npm-run-all": "^4.1.1",
     "nsp": "^2.8.1",
     "onchange": "^3.2.1",
-    "path": "^0.12.7",
```
Comment (biancadanforth) [inline package.json]: We use this devDependency in [run-firefox.js](https://github.com/mozilla/shield-studies-addon-template/blob/master/run-firefox.js#L13); can you add it back?

Comment (motin) [reply package.json]: That removal comes from https://github.com/mozilla/shield-studies-addon-template/pull/41 - comment that this dependency is used, close the pr and I will rebase this pr to not include this removal :)

File: addon/bootstrap.js (hunk 12 above)
Comment (motin) [reply addon/bootstrap.js]: Originally, I separated them for the ability to have to extension startup hooks - one for when the webextension had loaded asynchronously, and one at the end of startup() for privileged startup logic. Currently, there is only one startup method but it is anyway clearer to not have side-effects happen in a constructor event.

File: addon/bootstrap.js [hunk 14]
```diff
@@ -83,13 +83,21 @@ async function startup(addonData, reason) {
...
+  // log what the study variation and other info is.
+  log.debug(`info ${JSON.stringify(studyUtils.info())}`);
 
   // initiate the chrome-privileged part of the study add-on
   this.feature = new Feature(variation, studyUtils, REASONS[reason], log);
 
+  // if you have code to handle expiration / long-timers, it could go here
+  /*
+  if (this.feature.hasExpired()) {
+    // Please note that the general study expiration should probably be taken care of by Normandy.
```
Comment (biancadanforth) [inline addon/bootstrap.js]: This is news to me! I have been handling general study expiration in my addon code -- @mythmon , does Normandy handle this? What do you recommend study authors do re: general study expiration?

**Comment by motin** [comment]
> @motin That's a great point about the logging. Could you add a link to the pioneerUtils as a comment in that issue as a great example of how to do that?

Done

File: addon/bootstrap.js (hunk 14 above)
Comment (mythmon) [reply addon/bootstrap.js]: Normandy does not handle this, and I do not agree with this comment.

Normandy will handle removal of the add-on when the study is ended on the server, but that is not what this feature is generally used for. The first expiration timer that I'm aware of was designed to expire an experiment as a fail safe, in case Normandy could not be contacted, or someone forgot to disable the recipe on the server.

In that original case, the expiry timer was not a normal end to the study, but a failsafe in case something went wrong. As such, it should be implemented in a very reliable way by the add-on.

File: TESTPLAN.md [hunk 15]
```diff
@@ -1,117 +1,40 @@
-# Test Plan for the 57-Perception-Study Addon
+# Test plan for this add-on
 
-## User Experience / Functionality
+## Manual / QA TEST Instructions
+
+### Preparations
 
-During INSTALL ONLY users see:
+* Download a Release version of Firefox (Release is required for the recommendation heuristics to work)
```
Comment (biancadanforth) [inline TESTPLAN.md]: What do you mean by "recommendation heuristics"? Can you expand on that a little here?

Comment (motin) [reply TESTPLAN.md]: Sorry, a leftover from taar-v2, removing...

File: addon/bootstrap.js (hunk 14 above)
Comment (motin) [reply addon/bootstrap.js]: Thanks for clarifying @mythmon, adding this information as a comment in the template

File: package.json (hunk 13 above)
Comment (pdehaan) [reply package.json]: The [`path` module is the built in Node.js module](https://nodejs.org/dist/latest-v8.x/docs/api/path.html). Adding it in package.json includes a module from [npm](https://www.npmjs.com/package/path) &mdash; which could potentially be dangerous.

Comment (biancadanforth) [reply package.json]: Noted, and I have merged #41. Thank you Peter for clarifying!

File: addon/bootstrap.js [hunk 16]
```diff
@@ -119,22 +119,26 @@ function shutdown(addonData, reason) {
   // FRAGILE: handle uninstalls initiated by USER or by addon
   if (reason === REASONS.ADDON_UNINSTALL || reason === REASONS.ADDON_DISABLE) {
     log.debug("uninstall or disable");
+
```
Comment (biancadanforth) [inline addon/bootstrap.js]: Can you add a note here to say we check if `feature` exists because it's possible the study is shutting down before it has instantiated the feature. Ex: if the user is ineligible or if the study has expired.

Comment (motin) [reply addon/bootstrap.js]: Clarifying this

**Comment by motin** [comment]
Rebased on current master. 

They other PRs can be approved in this order: 
1. Close #35 (since it's change was already included in #41). 
2. Approve #39 and #30 in any order.

After that, I'll rebase this PR on top of the current master.

File: TESTPLAN.md [hunk 17]
```diff
@@ -123,8 +46,7 @@ All UI `shield-study` `study_state` sequences look like this:
     Test fails IF:
 
     - there is no bar.
-    - elements are not correct or are not displayed
-
+    - elements are not correct or are not displaye
```
Comment (biancadanforth) [inline TESTPLAN.md]: Nit: typo "displayed"

File: TESTPLAN.md [hunk 18]
```diff
@@ -163,72 +85,56 @@ All UI `shield-study` `study_state` sequences look like this:
...
+
+- prove the notification bar ui opens
+- *clicking on the left-most button presented*.
+- verifying that sent Telemetry is correct.
+
+Code at [/test/functional_test.js](/test/functional_test.js).
+
+### Note: checking "sent Telemetry is correct"
+
+* Open the Browser Console using Firefox's top menu at `Tools > Web Developer > Browser Console`. This will display Shield (loading/telemetry) log output from the add-on.
```
Comment (biancadanforth) [inline TESTPLAN.md]: Perhaps we can reference TELEMETRY.md and/or TESTPLAN.md here for how to verify "Telemetry is correct", since that's a nebulous phrase? In general I think we can expand on this, but that can be a separate PR later.

File: DEV.md [hunk 19]
```diff
@@ -14,54 +18,108 @@ npm run build
...
+
+See [TESTPLAN.md](./TESTPLAN.md) for more details on how to see this add-on in action and hot it is expected to behave.
+
+## Automated launch of Firefox with add-on installed
 
-This packages the add-on into `dist/linked-addon.xpi`. This file is the addon you load into Firefox.
+`$ npm run firefox` starts Firefox and automatically installs the add-on in a new profile and opens the browser console automatically.
 
-Note: `linked-addon.xpi` is a symbolic link to the extension's true XPI, which is named based on the study's unique addon ID specified in `package.json`.
+Note: This runs in a recently created profile. To have the study run despite the eligibility requirement of having at least 1 day old profiles, a config override is set in place to force the study to run.
```
Comment (biancadanforth) [inline DEV.md]: What's this note referring to? I'm not aware of any eligibility requirement for profile age. What is the config override pref?

File: docs/DEV.md [hunk 20]
```diff
@@ -0,0 +1,179 @@
...
+
+You can automatically build recent changes and package them into a `.xpi` by running the following from the top level directory:
+
+`$ npm run watch`
+
+Now, anytime a file is changed and saved, node will repackage the add-on. You must reload the add-on as before, or by clicking the "Reload" under the add-on in _about:debugging_. Note that a hard re-load is recommended to clear local storage. To do this, simply remove the add-on and reload as before.
+
+Note: This is currently only useful if you load the extension manually - it has no effect when running `npm run firefox`.
+
+## Directory Structure and Files
```
Comment (biancadanforth) [inline docs/DEV.md]: This file tree may need a refresh. For example, you removed `about.md`, `survival.md` and `tutorial.md` files, but they're still listed here.

File: .gitignore [hunk 21]
```diff
@@ -3,3 +3,4 @@ install.rdf
 *.xpi
 node_modules
 addon/StudyUtils.jsm
+addon/PioneerUtils.jsm
```
Comment (biancadanforth) [inline .gitignore]: Same as [this comment](https://github.com/mozilla/shield-studies-addon-template/pull/49/files#r170994130); consider removing.

File: addon/bootstrap.js [hunk 22]
```diff
@@ -1,166 +1,214 @@
...
-function shutdown(addonData, reason) {
-  log.debug("shutdown", REASONS[reason] || reason);
-  // FRAGILE: handle uninstalls initiated by USER or by addon
-  if (reason === REASONS.ADDON_UNINSTALL || reason === REASONS.ADDON_DISABLE) {
-    log.debug("uninstall or disable");
-    if (!studyUtils._isEnding) {
-      // we are the first 'uninstall' requestor => must be user action.
-      log.debug("probably: user requested shutdown");
-      studyUtils.endStudy({reason: "user-disable"});
+    // Expiration checks should be implemented in a very reliable way by the add-on since Normandy does not handle study expiration in a reliable manner
```
Comment (biancadanforth) [inline addon/bootstrap.js]: Thank you for updating this!

File: addon/Config.jsm [hunk 23]
```diff
@@ -72,6 +72,10 @@ var config = {
   // required LOG key
   "log": {
     // Fatal: 70, Error: 60, Warn: 50, Info: 40, Config: 30, Debug: 20, Trace: 10, All: -1,
+    "bootstrap": {
+      // Console.jsm uses "debug", whereas Log.jsm uses "Debug", *sigh*
```
Comment (biancadanforth) [inline addon/Config.jsm]: I'm glad you left in the *sigh* part of the comment. :)

File: docs/DEV.md [hunk 24]
```diff
@@ -0,0 +1,179 @@
...
+* Navigate to _about:config_ and set `extensions.legacy.enabled` to `true`. This permits the loading of the embedded Web Extension since new versions of Firefox are becoming restricted to pure Web Extensions only.
+* Navigate to _about:debugging_ in your URL bar
+* Select "Load Temporary Add-on"
+* Find and select the latest xpi file you just built.
+
+## Seeing the add-on in action
+
+To debug installation and loading of the add-on:
+
+* Navigate to _about:config_ and set `shield.testing.logging.level` to `10`. This permits shield-add-on log output in browser console
```
Comment (biancadanforth) [inline docs/DEV.md]: This currently doesn't do anything right now, so we should probably remove it until SSAU (utils) gets a log?

File: test/utils.js [hunk 25]
```diff
@@ -35,13 +35,16 @@ const FIREFOX_PREFERENCES = {
   // NECESSARY for all 57+ builds
   "extensions.legacy.enabled": true,
 
+  // Force variation for testing
+  "extensions.button_icon_preference.variation": "kittens",
```
Comment (biancadanforth) [inline test/utils.js]: If we're going to force the variation to "kittens", we should add a note in `DEV.md` and `TESTPLAN.md` that this pref can be used to force a variation (and similarly, removed from here to see how `weightedVariation` works from `Config.jsm` to select the treatment each time).

There should probably also be a comment in `bootstrap.js` where this pref is defined to explain this a bit more in the code.

File: .eslintignore [hunk 26]
```diff
@@ -1,7 +1,7 @@
 addon/StudyUtils.jsm
+addon/PioneerUtils.jsm
```
[the remark this answers is not in the document]
Comment (motin) [reply .eslintignore]: It will be added by anyone writing a Pioneer shield study, and they should still be able to use this template (me included :) )

File: test/utils.js (hunk 25 above)
Comment (motin) [reply test/utils.js]: Good suggestions. Done

File: docs/DEV.md (hunk 20 above)
Comment (motin) [reply docs/DEV.md]: Very true. Done

File: docs/DEV.md [hunk 27]
```diff
@@ -167,7 +167,7 @@ The web extension needs to be packaged together with a legacy add-on in order to
 
 It is recommended to build necessary logic and user interface using in the context of the webextension and communicate with the legacy add-on code through messaging whenever privileged access is required.
 
-For more information, see [./about.md]
+For more information, see <https://github.com/mozilla/shield-studies-addon-utils/blob/master/docs/engineering.md>.
```
Comment (biancadanforth) [inline docs/DEV.md]: This link doesn't work... I don't see an engineering.md in that repo.

File: .eslintignore (hunk 26 above)
[the remark this answers is not in the document]
Comment (biancadanforth) [reply .eslintignore]: Maybe just add a comment to explain what it's about? Something like:
`# This file is included for Pioneer Shield studies only; eventually it will be merged with StudyUtils.jsm`

File: addon/bootstrap.js [hunk 28]
```diff
@@ -45,7 +45,7 @@ XPCOMUtils.defineLazyModuleGetter(
 
 this.Bootstrap = {
   /**
-   * Preferences used by bootstrap.js
+   * Change this preference to test the add-on behavior in different study variations/branches with different (or remove it to test the automatic assigning of a study variation/branch).
```
Comment (biancadanforth) [inline addon/bootstrap.js]: Can you add a note where to find the branch strings to use? (i.e. in `weightedVariations` in `Config.jsm`)?

File: docs/DEV.md [hunk 29]
```diff
@@ -37,14 +37,6 @@ $ npm run build
 
 This packages the add-on into an xpi file which is stored in `dist/`. This file is what you load into Firefox.
 
-## General notes on Shield Study Engineering
```
Comment (biancadanforth) [inline docs/DEV.md]: I think this block provides some important context for devs on embedded webextensions; why did you remove it?

**Comment by biancadanforth** [comment]
@motin ; we're getting close! I have merged #39 and #30 , so please rebase your PR on top of those.

Before merging, I have a few clarifications/minor fix requests for you:
* Can you answer my question/comment re: [profile age](https://github.com/mozilla/shield-studies-addon-template/pull/49/commits/93e710998ec0169e6abaf936f3ee007b2603e63b#r170985485)? Not sure I am following there...
* This link to [engineering.md](https://github.com/mozilla/shield-studies-addon-template/pull/49/commits/e244b1df0cba67a36abb35fde0b9331e0b2b0165#r171397094) in the SSAU repo doesn't work. Not sure which document you are trying to reference, but can you fix the link?
* I think [this section](https://github.com/mozilla/shield-studies-addon-template/pull/49/commits/d874bb6268d9a33fe0ec9d45edde17e3b81bc8ec#r171400267) in DEV.md is actually pretty relevant (maybe it doesn't have the best wording, admittedly). My preference is to restore it. I'm open to others weighing in.

Also, if anyone else has any additional thoughts or concerns about merging this, please share them over the next few days. I'm especially interested to hear from @gregglind and @mythmon . In particular, Gregg, I wanted you to be aware of [this change](https://github.com/mozilla/shield-studies-addon-template/pull/49/commits/861a41a91f38055c99e8cf7a3597eb6d395045e4) to remove some legacy docs.

**Comment by gregglind** [comment]
I only care about docs changes that have useful faq or original research
stuff.

I am excited for the WEE future which cuts this template in half.

File: docs/DEV.md (hunk 27 above)
Comment (pdehaan) [reply docs/DEV.md]: I think docs/engineering.md was added in [PR #113](https://github.com/mozilla/shield-studies-addon-utils/pull/113/files#diff-06447854e2fbc871cd227e16c82c1962) in that repo.

File: docs/DEV.md (hunk 29 above)
Comment (motin) [reply docs/DEV.md]: I found there was two duplicate sections about this, so I removed one of them.

**Comment by motin** [comment]
>     Can you answer my question/comment re: profile age? Not sure I am following there...
That was a leftover comment from taar2 and is removed from this PR.
>     This link to engineering.md in the SSAU repo doesn't work. Not sure which document you are trying to reference, but can you fix the link?
It is referencing the "about.md" which gregg proposed to include in SSAU as "engineering.md" - a move and renaming I am fully behind. Thus, to fix the link, approve this PR: https://github.com/mozilla/shield-studies-addon-utils/pull/113 :)
>     I think this section in DEV.md is actually pretty relevant (maybe it doesn't have the best wording, admittedly). My preference is to restore it. I'm open to others weighing in.
I fully agree that it is relevant. However, it existed twice in the doc so I removed one of the dupes.

**Comment by biancadanforth** [comment]
Hi Freddy! I reviewed the [utils PR](https://github.com/mozilla/shield-studies-addon-utils/pull/113) you reference here to merge before merging this. I'll plan to wait until that's merged before a final review/merge of this PR.

File: docs/DEV.md [hunk 30]
```diff
@@ -0,0 +1,161 @@
...
+│   ├── install.rdf       # (derived from templates)
+│   ├── lib               # JSM (Firefox modules)
+│   │   └── Feature.jsm   # contains study-specific privileged code
+│   └── webextension      # study-specific embedded webextension
+│       ├── .eslintrc.json
+│       ├── background.js
+│       ├── icons
+│       │   ├── Anonymous-Lizard.svg
+│       │   ├── DogHazard1.svg
+│       │   ├── Grooming-Cat-Line-Art.svg
```
Comment (biancadanforth) [inline docs/DEV.md]: We probably want to remove these three from this [commit](https://github.com/mozilla/shield-studies-addon-template/pull/49/commits/ec44d6d53181735a715270a514c97f12a3094162) in the PR:
* Anonymous-Lizard.svg
* DogHazard1.svg
* Grooming-Cat-Line-Art.svg

**Comment by motin** [comment]
Rebased upon the latest master. Ready for merge.

**Comment by motin** [comment]
Rebased upon the latest master. Ready for merge.
